File: packages/server/cmd/renewlet/ai_models.go
package main

// ai_models.go 是 Docker/Go 运行面的 AI 模型列表代理。
//
// 该端点只在用户显式刷新时访问第三方 /models；Renewlet 发出的 API key 不入库、不回显，
// provider 原始响应只随当前认证错误返回，供设置页详情弹窗排查。
import (
	"context"
	"encoding/json"
	"errors"
	"fmt"
	"io"
	"net/http"
	"strings"
	"time"

	"github.com/pocketbase/pocketbase/core"
)

const (
	aiModelListTimeout       = 15 * time.Second
	aiModelListResponseBytes = 1 << 20
	aiModelListMaxModels     = 300
)

type aiModelListRequest struct {
	ProviderType string `json:"providerType"`
	BaseURL      string `json:"baseUrl"`
	APIKey       string `json:"apiKey"`
}

func (r *aiModelListRequest) Validate(locale appLocale) error {
	r.ProviderType = strings.TrimSpace(r.ProviderType)
	r.BaseURL = strings.TrimSpace(r.BaseURL)
	r.APIKey = strings.TrimSpace(r.APIKey)
	if !isValidAIRecognitionProviderType(r.ProviderType) {
		return errors.New(serverText(locale, "common.invalidRequestParameters"))
	}
	if r.BaseURL != "" && sanitizeAIRecognitionBaseURL(r.BaseURL) == "" {
		return errors.New(serverText(locale, "common.invalidRequestParameters"))
	}
	endpoint := resolveAIProviderEndpoint(aiRecognitionSettings{
		ProviderType: r.ProviderType,
		BaseURL:      r.BaseURL,
		APIKey:       r.APIKey,
	})
	if endpoint.BaseURLRequired && r.BaseURL == "" {
		return errors.New(serverText(locale, "aiRecognition.baseUrlRequired"))
	}
	if endpoint.APIKeyRequired && r.APIKey == "" {
		return errors.New(serverText(locale, "aiRecognition.apiKeyRequired"))
	}
	return nil
}

type aiModelCapabilities struct {
	TextInput        *bool `json:"textInput"`
	ImageInput       *bool `json:"imageInput"`
	StructuredOutput *bool `json:"structuredOutput"`
	Thinking         *bool `json:"thinking"`
}

type aiModelListItem struct {
	ID               string              `json:"id"`
	DisplayName      *string             `json:"displayName"`
	CreatedAt        *string             `json:"createdAt"`
	OwnedBy          *string             `json:"ownedBy"`
	InputTokenLimit  *int                `json:"inputTokenLimit"`
	OutputTokenLimit *int                `json:"outputTokenLimit"`
	Capabilities     aiModelCapabilities `json:"capabilities"`
}

type aiModelListResponse struct {
	ProviderType      string            `json:"providerType"`
	TransportProtocol string            `json:"transportProtocol"`
	Models            []aiModelListItem `json:"models"`
	Truncated         bool              `json:"truncated"`
}

type aiModelListErrorResponse struct {
	Message string                  `json:"message"`
	Code    string                  `json:"code"`
	Details aiModelListErrorDetails `json:"details"`
}

type aiModelListErrorDetails struct {
	Reason           string              `json:"reason"`
	ProviderMessage  *string             `json:"providerMessage"`
	ProviderResponse *aiProviderResponse `json:"providerResponse,omitempty"`
}

type aiModelListEndpoint struct {
	URL               string
	Headers           http.Header
	ModelListShape    string
	ProviderType      string
	TransportProtocol string
}

type aiModelListHTTPError struct {
	status           int
	code             string
	reason           string
	message          *string
	providerResponse *aiProviderResponse
}

func (e *aiModelListHTTPError) Error() string {
	if e.message != nil {
		return *e.message
	}
	return e.reason
}

var aiModelListHTTPClient = &http.Client{Timeout: aiModelListTimeout}

func handleAIModelsList(app core.App, e *core.RequestEvent) error {
	locale := requestLocale(e.Request)
	if e.Auth == nil {
		return e.UnauthorizedError(serverText(locale, "auth.loginRequired"), nil)
	}
	// 请求体携带临时 provider 配置，必须保持严格 JSON，避免未知字段变成隐式模型代理配置。
	body, err := decodeStrictJSON[aiModelListRequest](e.Request, locale)
	if err != nil {
		return e.BadRequestError(validationErrorMessage(locale, "common.invalidRequestBody", err), err)
	}
	response, err := listAIModels(e.Request.Context(), body, locale)
	if err != nil {
		var httpErr *aiModelListHTTPError
		if errors.As(err, &httpErr) {
			return e.JSON(httpErr.status, aiModelListErrorResponse{
				Message: serverText(locale, "aiRecognition.modelListFailed"),
				Code:    httpErr.code,
				Details: aiModelListErrorDetails{
					Reason:           httpErr.reason,
					ProviderMessage:  httpErr.message,
					ProviderResponse: httpErr.providerResponse,
				},
			})
		}
		return e.JSON(http.StatusBadRequest, aiModelListErrorResponse{
			Message: serverText(locale, "aiRecognition.modelListFailed"),
			Code:    "AI_MODEL_LIST_FAILED",
			Details: aiModelListErrorDetails{Reason: "provider_failed", ProviderMessage: optionalAIProviderBody(err.Error())},
		})
	}
	return e.JSON(http.StatusOK, response)
}

func listAIModels(ctx context.Context, input aiModelListRequest, locale appLocale) (aiModelListResponse, error) {
	endpoint, err := buildAIModelListEndpoint(input)
	if err != nil {
		return aiModelListResponse{}, err
	}
	raw, err := fetchAIModelListJSON(ctx, endpoint, locale)
	if err != nil {
		return aiModelListResponse{}, err
	}
	models := normalizeAIModelList(endpoint.ModelListShape, raw)
	truncated := len(models) > aiModelListMaxModels
	if truncated {
		// 截断只影响下拉候选展示；用户仍可手动输入模型 ID，避免超大 provider 响应拖垮页面。
		models = models[:aiModelListMaxModels]
	}
	return aiModelListResponse{ProviderType: endpoint.ProviderType, TransportProtocol: endpoint.TransportProtocol, Models: models, Truncated: truncated}, nil
}

func buildAIModelListEndpoint(input aiModelListRequest) (aiModelListEndpoint, error) {
	headers := http.Header{"Accept": []string{"application/json"}}
	if !isValidAIRecognitionProviderType(input.ProviderType) {
		return aiModelListEndpoint{}, errAIRecognitionProviderInvalid
	}
	endpoint := resolveAIProviderEndpoint(aiRecognitionSettings{
		ProviderType: input.ProviderType,
		BaseURL:      input.BaseURL,
		APIKey:       input.APIKey,
	})
	for key, values := range endpoint.Headers {
		for _, value := range values {
			headers.Add(key, value)
		}
	}
	return aiModelListEndpoint{
		URL:               endpoint.ModelsURL,
		Headers:           headers,
		ModelListShape:    endpoint.ModelListShape,
		ProviderType:      endpoint.ProviderType,
		TransportProtocol: endpoint.TransportProtocol,
	}, nil
}

func fetchAIModelListJSON(ctx context.Context, endpoint aiModelListEndpoint, locale appLocale) (map[string]interface{}, error) {
	ctx, cancel := context.WithTimeout(ctx, aiModelListTimeout)
	defer cancel()
	request, err := http.NewRequestWithContext(ctx, http.MethodGet, endpoint.URL, nil)
	if err != nil {
		return nil, err
	}
	for key, values := range endpoint.Headers {
		for _, value := range values {
			request.Header.Add(key, value)
		}
	}
	// 模型列表刷新是用户显式动作；第三方凭证只参与服务端代理请求，不写库、不回传浏览器。
	response, err := aiModelListHTTPClient.Do(request)
	if err != nil {
		if errors.Is(ctx.Err(), context.DeadlineExceeded) || errors.Is(err, context.DeadlineExceeded) {
			return nil, &aiModelListHTTPError{status: http.StatusRequestTimeout, code: "AI_MODEL_LIST_TIMEOUT", reason: "timeout"}
		}
		return nil, err
	}
	defer func() { _ = response.Body.Close() }()
	body, err := readAIModelListResponseBody(response.Body, locale)
	if err != nil {
		return nil, err
	}
	if response.StatusCode < 200 || response.StatusCode >= 300 {
		message := string(body)
		return nil, &aiModelListHTTPError{
			status:           response.StatusCode,
			code:             "AI_MODEL_LIST_FAILED",
			reason:           fmt.Sprintf("http_%d", response.StatusCode),
			message:          optionalAIProviderBody(message),
			providerResponse: aiProviderResponseFromHTTPResponse(response, message),
		}
	}
	var raw map[string]interface{}
	if err := json.Unmarshal(body, &raw); err != nil {
		message := string(body)
		return nil, &aiModelListHTTPError{
			status:           http.StatusBadRequest,
			code:             "AI_MODEL_LIST_INVALID_JSON",
			reason:           "invalid_json",
			message:          optionalAIProviderBody(message),
			providerResponse: aiProviderResponseFromHTTPResponse(response, message),
		}
	}
	return raw, nil
}

func readAIModelListResponseBody(reader io.Reader, locale appLocale) ([]byte, error) {
	data, err := io.ReadAll(io.LimitReader(reader, aiModelListResponseBytes+1))
	if err != nil {
		return nil, err
	}
	if len(data) > aiModelListResponseBytes {
		// 第三方错误页可能远大于正常模型列表；限制响应体能防止服务端代理变成内存放大器。
		return nil, &aiModelListHTTPError{
			status:  http.StatusRequestEntityTooLarge,
			code:    "AI_MODEL_LIST_RESPONSE_TOO_LARGE",
			reason:  "response_too_large",
			message: optionalAIProviderBody(serverText(locale, "common.requestBodyTooLarge")),
		}
	}
	return data, nil
}

func normalizeAIModelList(shape string, raw map[string]interface{}) []aiModelListItem {
	var models []aiModelListItem
	switch shape {
	case "gemini":
		models = normalizeGeminiModelList(raw)
	case "anthropic":
		models = normalizeAnthropicModelList(raw)
	default:
		models = normalizeOpenAIShapeModelList(raw)
	}
	return dedupeAIModelList(models)
}

func normalizeOpenAIShapeModelList(raw map[string]interface{}) []aiModelListItem {
	items := arrayFromMap(raw, "data")
	models := make([]aiModelListItem, 0, len(items))
	for _, item := range items {
		record := mapFromValue(item)
		id := stringFromMap(record, "id")
		if id == "" {
			continue
		}
		models = append(models, newAIModelListItem(aiModelListItem{
			ID:          id,
			DisplayName: optionalFirstNonBlank(stringFromMap(record, "display_name"), stringFromMap(record, "displayName")),
			CreatedAt:   optionalEpochSecondsISO(numberFromMap(record, "created")),
			OwnedBy:     optionalFirstNonBlank(stringFromMap(record, "owned_by"), stringFromMap(record, "ownedBy")),
		}))
	}
	return models
}

func normalizeGeminiModelList(raw map[string]interface{}) []aiModelListItem {
	items := arrayFromMap(raw, "models")
	models := make([]aiModelListItem, 0, len(items))
	for _, item := range items {
		record := mapFromValue(item)
		methods := stringSliceFromMap(record, "supportedGenerationMethods")
		if len(methods) > 0 && !containsString(methods, "generateContent") {
			continue
		}
		id := firstNonBlank(stringFromMap(record, "baseModelId"), strings.TrimPrefix(stringFromMap(record, "name"), "models/"))
		if id == "" {
			continue
		}
		textInput := true
		model := newAIModelListItem(aiModelListItem{
			ID:               id,
			DisplayName:      optionalFirstNonBlank(stringFromMap(record, "displayName")),
			InputTokenLimit:  optionalPositiveInt(numberFromMap(record, "inputTokenLimit")),
			OutputTokenLimit: optionalPositiveInt(numberFromMap(record, "outputTokenLimit")),
			Capabilities: aiModelCapabilities{
				TextInput: &textInput,
				Thinking:  optionalThinkingCapability(record["thinking"]),
			},
		})
		models = append(models, model)
	}
	return models
}

func normalizeAnthropicModelList(raw map[string]interface{}) []aiModelListItem {
	items := arrayFromMap(raw, "data")
	models := make([]aiModelListItem, 0, len(items))
	for _, item := range items {
		record := mapFromValue(item)
		id := stringFromMap(record, "id")
		if id == "" {
			continue
		}
		capabilities := mapFromValue(record["capabilities"])
		models = append(models, newAIModelListItem(aiModelListItem{
			ID:          id,
			DisplayName: optionalFirstNonBlank(stringFromMap(record, "display_name"), stringFromMap(record, "displayName")),
			CreatedAt:   optionalFirstNonBlank(stringFromMap(record, "created_at"), stringFromMap(record, "createdAt")),
			OwnedBy:     optionalFirstNonBlank(stringFromMap(record, "type")),
			Capabilities: aiModelCapabilities{
				TextInput:  optionalBoolFromMap(capabilities, "text"),
				ImageInput: optionalBoolFromMap(capabilities, "vision"),
				Thinking:   optionalBoolFromMap(capabilities, "thinking"),
			},
		}))
	}
	return models
}

func newAIModelListItem(item aiModelListItem) aiModelListItem {
	item.ID = strings.TrimSpace(item.ID)
	item.Capabilities = aiModelCapabilities{
		TextInput:        item.Capabilities.TextInput,
		ImageInput:       item.Capabilities.ImageInput,
		StructuredOutput: item.Capabilities.StructuredOutput,
		Thinking:         item.Capabilities.Thinking,
	}
	return item
}

func dedupeAIModelList(models []aiModelListItem) []aiModelListItem {
	out := make([]aiModelListItem, 0, len(models))
	seen := map[string]struct{}{}
	for _, model := range models {
		key := strings.ToLower(model.ID)
		if key == "" {
			continue
		}
		if _, ok := seen[key]; ok {
			continue
		}
		seen[key] = struct{}{}
		out = append(out, model)
	}
	return out
}

func mapFromValue(value interface{}) map[string]interface{} {
	if data, ok := value.(map[string]interface{}); ok {
		return data
	}
	return map[string]interface{}{}
}

func arrayFromMap(value map[string]interface{}, key string) []interface{} {
	if data, ok := value[key].([]interface{}); ok {
		return data
	}
	return nil
}

func stringFromMap(value map[string]interface{}, key string) string {
	if data, ok := value[key].(string); ok {
		return strings.TrimSpace(data)
	}
	return ""
}

func numberFromMap(value map[string]interface{}, key string) float64 {
	if data, ok := value[key].(float64); ok {
		return data
	}
	return 0
}

func stringSliceFromMap(value map[string]interface{}, key string) []string {
	raw, ok := value[key].([]interface{})
	if !ok {
		return nil
	}
	out := make([]string, 0, len(raw))
	for _, item := range raw {
		if text, ok := item.(string); ok && strings.TrimSpace(text) != "" {
			out = append(out, strings.TrimSpace(text))
		}
	}
	return out
}

func optionalBoolFromMap(value map[string]interface{}, key string) *bool {
	if data, ok := value[key].(bool); ok {
		return &data
	}
	return nil
}

func optionalThinkingCapability(value interface{}) *bool {
	if data, ok := value.(bool); ok {
		return &data
	}
	if value != nil {
		data := true
		return &data
	}
	return nil
}

func optionalPositiveInt(value float64) *int {
	if value <= 0 {
		return nil
	}
	out := int(value)
	return &out
}

func optionalEpochSecondsISO(value float64) *string {
	if value <= 0 {
		return nil
	}
	text := time.Unix(int64(value), 0).UTC().Format(time.RFC3339)
	return &text
}

func optionalFirstNonBlank(values ...string) *string {
	value := firstNonBlank(values...)
	if value == "" {
		return nil
	}
	return &value
}

func containsString(values []string, target string) bool {
	for _, value := range values {
		if value == target {
			return true
		}
	}
	return false
}

File: apps/worker/src/ai-models.ts
/**
 * Cloudflare AI 模型列表代理只服务“用户主动刷新”场景。
 *
 * 这里复用 shared provider endpoint 作为请求事实源，Worker 只负责认证、超时、响应限额和错误回显；
 * 模型候选不会入库，避免把第三方账号能力或请求 API key 变成 Renewlet 的持久化数据。
 */
import {
  AI_RECOGNITION_MAX_MODEL_LIST_MODELS,
  aiModelListErrorDetailsSchema,
  aiModelListPayloadSchema,
  aiModelListRequestSchema,
  type AiModelListItem,
  type AiModelListRequest,
  type AiRecognitionTransportProtocol,
} from "@renewlet/shared/schemas/ai-recognition";
import { resolveAIProviderEndpoint, type AIModelListResponseShape } from "@renewlet/shared/ai-provider-endpoints";
import { requireAuth } from "./auth";
import { HttpError, readJson, requestLocale, successJson } from "./http";
import { serverText, type AppLocale } from "./server-i18n";
import type { Env } from "./types";
import { providerResponseFromFetchResponse } from "./ai-provider-response";
import type { UpstreamProviderResponse as AiProviderResponse } from "./upstream-response";
import { UpstreamRequestError, sendUpstreamRequest } from "./upstream-http";

const AI_MODEL_LIST_TIMEOUT_MS = 15_000;
const AI_MODEL_LIST_RESPONSE_BYTES = 1 << 20;

type ModelListEndpoint = {
  url: string;
  headers: Headers;
  secrets: readonly string[];
  modelListShape: AIModelListResponseShape;
  providerType: AiModelListRequest["providerType"];
  transportProtocol: AiRecognitionTransportProtocol;
};

type NormalizedModelList = {
  models: AiModelListItem[];
  truncated: boolean;
};

type AIModelListResponseText = {
  text: string;
  truncated: boolean;
};

/**
 * listAIModels 是用户显式刷新模型列表的认证代理。
 *
 * 请求 API key 只在 Worker 内用于第三方 `/models` 请求，不入库、不回显，也不由前端直连 provider。
 */
export async function listAIModels(request: Request, env: Env): Promise<Response> {
  const locale = requestLocale(request);
  await requireAuth(request, env);
  const body = await readJson(request, aiModelListRequestSchema, locale);
  const input = normalizeAIModelListRequest(body);
  assertAIModelListRequest(input, locale);

  try {
    const endpoint = buildAIModelListEndpoint(input);
    const raw = await fetchAIModelListJSON(endpoint, locale);
    const normalized = normalizeAIModelList(endpoint.modelListShape, raw);
    return successJson(aiModelListPayloadSchema.parse({
      providerType: endpoint.providerType,
      transportProtocol: endpoint.transportProtocol,
      models: normalized.models,
      truncated: normalized.truncated,
    }));
  } catch (error) {
    if (error instanceof HttpError) throw error;
    throw new HttpError(
      400,
      serverText(locale, "aiRecognition.modelListFailed"),
      "AI_MODEL_LIST_FAILED",
      aiModelListErrorDetails("provider_failed", error),
    );
  }
}

function normalizeAIModelListRequest(input: AiModelListRequest): AiModelListRequest {
  return {
    providerType: input.providerType,
    baseUrl: input.baseUrl.trim(),
    apiKey: input.apiKey.trim(),
  };
}

function assertAIModelListRequest(input: AiModelListRequest, locale: AppLocale): void {
  const endpoint = resolveAIProviderEndpoint(input);
  if (endpoint.baseUrlRequired && !input.baseUrl) {
    throw new HttpError(400, serverText(locale, "aiRecognition.baseUrlRequired"), "AI_MODEL_LIST_BASE_URL_REQUIRED");
  }
  if (endpoint.apiKeyRequired && !input.apiKey) {
    throw new HttpError(400, serverText(locale, "aiRecognition.apiKeyRequired"), "AI_MODEL_LIST_API_KEY_REQUIRED");
  }
}

function buildAIModelListEndpoint(input: AiModelListRequest): ModelListEndpoint {
  const headers = new Headers({ accept: "application/json" });
  const endpoint = resolveAIProviderEndpoint(input);
  // 鉴权头由 shared resolver 按协议生成；Worker 只补模型列表代理自己的 Accept。
  for (const [key, value] of Object.entries(endpoint.authHeaders)) {
    headers.set(key, value);
  }
  return {
    url: endpoint.modelsUrl,
    headers,
    secrets: input.apiKey ? [input.apiKey] : [],
    modelListShape: endpoint.modelListShape,
    providerType: endpoint.providerType,
    transportProtocol: endpoint.transportProtocol,
  };
}

async function fetchAIModelListJSON(endpoint: ModelListEndpoint, locale: AppLocale): Promise<unknown> {
  let response: Response;
  try {
    // 这是用户显式点击刷新后才触发的第三方请求；请求 API key 只在 Worker 侧使用，不回显、不持久化。
    response = await sendUpstreamRequest(endpoint.url, {
      method: "GET",
      headers: endpoint.headers,
    }, {
      provider: `${endpoint.providerType} models`,
      secrets: endpoint.secrets,
      timeoutMs: AI_MODEL_LIST_TIMEOUT_MS,
    });
  } catch (error) {
    if (error instanceof UpstreamRequestError && error.timedOut) {
      throw new HttpError(
        408,
        serverText(locale, "aiRecognition.modelListTimeout"),
        "AI_MODEL_LIST_TIMEOUT",
        aiModelListErrorDetails("timeout", error),
      );
    }
    throw error;
  }

  const body = await readAIModelListResponseText(response, locale);
  if (!response.ok) {
    throw new HttpError(
      response.status,
      serverText(locale, "aiRecognition.modelListFailed"),
      "AI_MODEL_LIST_FAILED",
      aiModelListErrorDetails(`http_${response.status}`, body.text, providerResponseFromFetchResponse(response, body.text, body.truncated, endpoint.secrets)),
    );
  }

  try {
    return JSON.parse(body.text) as unknown;
  } catch {
    throw new HttpError(
      400,
      serverText(locale, "aiRecognition.modelListFailed"),
      "AI_MODEL_LIST_INVALID_JSON",
      aiModelListErrorDetails("invalid_json", body.text, providerResponseFromFetchResponse(response, body.text, body.truncated, endpoint.secrets)),
    );
  }
}

async function readAIModelListResponseText(response: Response, locale: AppLocale): Promise<AIModelListResponseText> {
  const reader = response.body?.getReader();
  if (!reader) return { text: "", truncated: false };
  const decoder = new TextDecoder();
  let total = 0;
  let text = "";
  for (;;) {
    const { done, value } = await reader.read();
    if (done) break;
    total += value.byteLength;
    if (total > AI_MODEL_LIST_RESPONSE_BYTES) {
      // 第三方模型接口不应返回海量正文；超限直接取消 reader，避免 Worker 内存被 provider 错误页放大。
      await reader.cancel().catch(() => undefined);
      throw new HttpError(
        413,
        serverText(locale, "common.requestBodyTooLarge"),
        "AI_MODEL_LIST_RESPONSE_TOO_LARGE",
        aiModelListErrorDetails("response_too_large", null),
      );
    }
    text += decoder.decode(value, { stream: true });
  }
  return { text: text + decoder.decode(), truncated: false };
}

function normalizeAIModelList(shape: AIModelListResponseShape, raw: unknown): NormalizedModelList {
  const models = shape === "gemini"
    ? normalizeGeminiModels(raw)
    : shape === "anthropic"
      ? normalizeAnthropicModels(raw)
      : normalizeOpenAIShapeModels(raw);
  const deduped = dedupeModels(models);
  return {
    // 截断只影响候选展示，用户仍可手输模型 ID；列表不能成为超大 provider 响应的搬运通道。
    models: deduped.slice(0, AI_RECOGNITION_MAX_MODEL_LIST_MODELS),
    truncated: deduped.length > AI_RECOGNITION_MAX_MODEL_LIST_MODELS,
  };
}

function normalizeOpenAIShapeModels(raw: unknown): AiModelListItem[] {
  const data = arrayField(record(raw), "data");
  return data.map((item) => {
    const model = record(item);
    const id = stringField(model, "id");
    if (!id) return null;
    return modelItem({
      id,
      displayName: stringField(model, "display_name") || stringField(model, "displayName"),
      createdAt: epochSecondsToISO(numberField(model, "created")),
      ownedBy: stringField(model, "owned_by") || stringField(model, "ownedBy"),
    });
  }).filter((item): item is AiModelListItem => item !== null);
}

function normalizeGeminiModels(raw: unknown): AiModelListItem[] {
  const models = arrayField(record(raw), "models");
  return models.map((item) => {
    const model = record(item);
    const methods = stringArrayField(model, "supportedGenerationMethods");
    if (methods.length > 0 && !methods.includes("generateContent")) return null;
    const name = stringField(model, "name");
    const id = stringField(model, "baseModelId") || stripGeminiModelPrefix(name);
    if (!id) return null;
    const thinking = model["thinking"];
    return modelItem({
      id,
      displayName: stringField(model, "displayName"),
      inputTokenLimit: numberField(model, "inputTokenLimit"),
      outputTokenLimit: numberField(model, "outputTokenLimit"),
      capabilities: {
        textInput: true,
        imageInput: null,
        structuredOutput: null,
        thinking: typeof thinking === "boolean" ? thinking : thinking == null ? null : true,
      },
    });
  }).filter((item): item is AiModelListItem => item !== null);
}

function normalizeAnthropicModels(raw: unknown): AiModelListItem[] {
  const data = arrayField(record(raw), "data");
  return data.map((item) => {
    const model = record(item);
    const id = stringField(model, "id");
    if (!id) return null;
    const capabilities = record(model["capabilities"]);
    return modelItem({
      id,
      displayName: stringField(model, "display_name") || stringField(model, "displayName"),
      createdAt: stringField(model, "created_at") || stringField(model, "createdAt"),
      ownedBy: stringField(model, "type"),
      capabilities: {
        textInput: booleanField(capabilities, "text") ?? null,
        imageInput: booleanField(capabilities, "vision") ?? null,
        structuredOutput: null,
        thinking: booleanField(capabilities, "thinking") ?? null,
      },
    });
  }).filter((item): item is AiModelListItem => item !== null);
}

function modelItem(input: {
  id: string;
  displayName?: string | null;
  createdAt?: string | null;
  ownedBy?: string | null;
  inputTokenLimit?: number | null;
  outputTokenLimit?: number | null;
  capabilities?: Partial<AiModelListItem["capabilities"]>;
}): AiModelListItem {
  return {
    id: input.id,
    displayName: input.displayName || null,
    createdAt: input.createdAt || null,
    ownedBy: input.ownedBy || null,
    inputTokenLimit: input.inputTokenLimit ?? null,
    outputTokenLimit: input.outputTokenLimit ?? null,
    capabilities: {
      textInput: input.capabilities?.textInput ?? null,
      imageInput: input.capabilities?.imageInput ?? null,
      structuredOutput: input.capabilities?.structuredOutput ?? null,
      thinking: input.capabilities?.thinking ?? null,
    },
  };
}

function dedupeModels(models: readonly AiModelListItem[]): AiModelListItem[] {
  const out: AiModelListItem[] = [];
  const seen = new Set<string>();
  for (const model of models) {
    const key = model.id.toLowerCase();
    if (seen.has(key)) continue;
    seen.add(key);
    out.push(model);
  }
  return out;
}

function record(value: unknown): Record<string, unknown> {
  return value && typeof value === "object" && !Array.isArray(value) ? value as Record<string, unknown> : {};
}

function arrayField(value: Record<string, unknown>, key: string): unknown[] {
  const field = value[key];
  return Array.isArray(field) ? field : [];
}

function stringField(value: Record<string, unknown>, key: string): string | null {
  const field = value[key];
  return typeof field === "string" && field.trim() ? field.trim() : null;
}

function stringArrayField(value: Record<string, unknown>, key: string): string[] {
  const field = value[key];
  return Array.isArray(field) ? field.filter((item): item is string => typeof item === "string").map((item) => item.trim()).filter(Boolean) : [];
}

function numberField(value: Record<string, unknown>, key: string): number | null {
  const field = value[key];
  return typeof field === "number" && Number.isFinite(field) ? Math.trunc(field) : null;
}

function booleanField(value: Record<string, unknown>, key: string): boolean | null {
  const field = value[key];
  return typeof field === "boolean" ? field : null;
}

function epochSecondsToISO(value: number | null): string | null {
  return value && value > 0 ? new Date(value * 1000).toISOString() : null;
}

function stripGeminiModelPrefix(value: string | null): string | null {
  if (!value) return null;
  return value.replace(/^models\//, "");
}

function aiModelListErrorDetails(reason: string, providerMessage: unknown, providerResponse: AiProviderResponse | null = null) {
  const message = providerMessage instanceof Error
    ? providerMessage.message
    : typeof providerMessage === "string"
      ? providerMessage
      : null;
  return aiModelListErrorDetailsSchema.parse({
    rawResponseText: providerResponse?.body || message || reason,
  });
}
